Feature/pinhole camera

julsanmer opened the pull request:
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? How are your
commits organized? -->
Creation of a pinholeCamera module to generate landmarks-based measurements around an asteroid (but not limited to). An integrated example script scenarioSmallBodyLandmarks.py is also added.

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? If you didn't 
add or update any tests justify this choice. -->
There is a unit test for this module, as well as an integrated example scenario that uses the pinhole camera module.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and 
completeness? -->
The module RST is included, and the example scenario discusses its use.

## Future work
<!-- What next steps can we anticipate from here, if any? -->
None.

patkenneally commented on src/simulation/navigation/pinholeCamera/pinholeCamera.cpp:
```diff
@@ -0,0 +1,328 @@
+        flagLighting = this->checkLighting(nLmk);
+        
+        /* Add pixel */
+        if (flagFOV and flagLighting){
```
If I'm not mistaken we don't enable the the `/permissive` or `/Za` flags on windows and are unlikely to at this moment. So you could fix you windows CI build by not using the `and` keyword, but rather `&&`.

julsanmer replied:
Thank you for signalling the issue. Will try to see if that fixes it as there are multiple complaints in the windows build

schaubh commented on src/simulation/navigation/pinholeCamera/pinholeCamera.h:
```diff
@@ -0,0 +1,100 @@
+    Eigen::MatrixXi pixelLmk; //!< [-] pixels for landmarks
+        
+    /* Landmark distribution */
+    int n; //!< [-] number of landmarks
```
shouldn't this be a private variable?  It is set in reset by your module, not the use.  Please check that only variables are public that should be set by the user or are need to be accessible outside of this class.

julsanmer replied:
Yes, it could be. I see no issues to have this one private. I may push this a little bit oftentimes as I like to access them to have the number at hand (e.g. field of view angles are computed internally but public)

schaubh commented on src/simulation/navigation/pinholeCamera/pinholeCamera.cpp:
```diff
@@ -0,0 +1,328 @@
+void PinholeCamera::loopLandmarks()
+{
+    /* Define variables */
+    Eigen::Vector3d rLmk, nLmk;
```
add comments defining what these variable are

julsanmer replied:
Added

schaubh commented on src/simulation/navigation/pinholeCamera/pinholeCamera.cpp:
```diff
@@ -0,0 +1,328 @@
+*/
+bool PinholeCamera::checkFOV(Eigen::Vector2i pLmk, Eigen::Vector3d rLmk, Eigen::Vector3d nLmk){
+    /* Compute focal direction in P frame */
+    Eigen::Vector3d eC_P, eC_C;
```
define what these variables are

julsanmer replied:
Done. Thanks for pointing this, I've rearranged these variables in the module

schaubh commented on src/simulation/navigation/pinholeCamera/pinholeCamera.cpp:
```diff
@@ -0,0 +1,328 @@
+    eC_P = (this->dcm_CB * this->dcm_BP).transpose()*eC_C;
+    
+    /* Define unit-vector from landmark to spacecraft */
+    Eigen::Vector3d dr, eLmk_P;
```
Please ensure all variables in this file have comments that explain what they are.

julsanmer replied:
Done

schaubh commented on src/architecture/msgPayloadDefC/AccessMsgPayload.h:
```diff
@@ -33,6 +33,8 @@ typedef struct {
     double r_BL_L[3]; //!<[m] Spacecraft position relative to the groundLocation in the SEZ frame.
     double v_BL_L[3]; //!<[m/s] SEZ relative time derivative of r_BL vector in SEZ vector components.
+    
+    int pL[2]; //!< [-] pixel of the location as view from pinhole camera
```
Mm, I don't see why this pixel info would be in the access msg?  I prefer that this pixel info is put into its own message type.

julsanmer replied:
Yeah, it was 50-50 decision to me, glad that you've made it. I've created a separate LandmarkMsgPayload.h message and reverted the access one.

schaubh commented on examples/scenarioSmallBodyLandmarks.py:
```diff
@@ -0,0 +1,571 @@
+#
+#  ISC License
+#
+#  Copyright (c) 2016, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
```
update 2016 to be the current year. :-)

julsanmer replied:
:) who's gonna trust people doing space software that does not know what year it is

schaubh commented on examples/scenarioSmallBodyLandmarks.py:
```diff
@@ -0,0 +1,571 @@
+#
+
+r"""
+Overview
```
This script doesn't run on my machine.  I'm getting error messages.

julsanmer replied:
My bad. I did a modification to gravCoeffOps.py to support another type of polyhedron file (as it is the one used in the script) but forgot to commit it. Being the landmark distribution based on the previous file and it does not load properly, god knows what is happening

I've committed the my change to gravCoeffOps.py, let me know if there is more trouble

schaubh replied:
Cool, the scenario scripts now runs.  Nice job with the 3D asteroid figures showing what landmarks are visible at two different time frames.

schaubh commented on examples/scenarioSmallBodyLandmarks.py:
```diff
@@ -0,0 +1,571 @@
+
+r"""
+Overview
+--------
```
This scenario script does not show up in the example scripts web page.  You have to add this to `basilisk/examples/_default.rst`.

julsanmer replied:
Yeah, forgot this as well as the block diagram image.

Added.

schaubh commented on examples/scenarioSmallBodyLandmarks.py:
```diff
@@ -0,0 +1,571 @@
+    plt.ylabel('Orientation Illustration')
+
+
+def run(show_plots, useAltBodyFrame=True, useBatch=False, useGravPoly=False):
```
Your run() argument takes several optional arguments.  But, you include this scenario in `test_scenarioTest.py` which only calls run() without arguments.  Thus, many versions of this code don't get tested.  Please write a dedicated `test_scenarioSmallBodyLandmark.py` file that parametrizes these options to ensure you run all combinations of this scenario.

julsanmer replied:
Thanks for clarifying this. I've tailored the arguments for that purpose and created a test_scenarioSmallBodyLandmark.py. It checks plots and that the Simulation and method provides same results

schaubh reviewed (changes requested):
thanks for making this cool contribution.  This will be very useful.  I've marked several items that need to be addressed.  I can't currently run your scenario script.  Thanks for looking into these questions and comments.  I'll then do another, more detailed review.  

I did build the documentation.  Nice work there documenting the module and the example script.  There are RST errors right now as it can't import the example script images as my system can't run the example script.

julsanmer commented:
> thanks for making this cool contribution. This will be very useful. I've marked several items that need to be addressed. I can't currently run your scenario script. Thanks for looking into these questions and comments. I'll then do another, more detailed review.
> 
> I did build the documentation. Nice work there documenting the module and the example script. There are RST errors right now as it can't import the example script images as my system can't run the example script.

Thank you for reviewing this and providing feedback Dr. Schaub. I've responded to your comments in the two new separate commits (one for the module issues and the other for the example script)

schaubh reviewed (changes requested):
Howdy Juan, nice changes.  The scripts build now, changes look good, unit tests do what they need, documentation builds without issues.  I still need
- rebase on develop
- add release notes to docs/Support/bskReleaseNotes.rst
- finally, please clean up this branch by merging the commits that fix earlier things into the original commits.  We should just see the final clean commits, not the various fixes.  If you have multiple fixes in one commit, please break up that commit and apply the fixes to the original separate commits.

julsanmer commented:
> Howdy Juan, nice changes. The scripts build now, changes look good, unit tests do what they need, documentation builds without issues. I still need
> 
> * rebase on develop
> * add release notes to docs/Support/bskReleaseNotes.rst
> * finally, please clean up this branch by merging the commits that fix earlier things into the original commits.  We should just see the final clean commits, not the various fixes.  If you have multiple fixes in one commit, please break up that commit and apply the fixes to the original separate commits.

Hello Dr. Schaub. I've rebased, added release notes and refactored the commits. Hope to not have break anything again, everything runs ok at my side and CI builds seems ok as well. Let me know if you need anything more and thank you for the detailed explanations.
